src/Compilers/CSharp/Portable/Symbols/UpdatedContainingSymbolLocal.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+#nullable enable
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal sealed class UpdatedContainingSymbolAndNullableAnnotationLocal : LocalSymbol
+    {
+        internal UpdatedContainingSymbolAndNullableAnnotationLocal(SourceLocalSymbol underlyingLocal, Symbol updatedContainingSymbol, TypeWithAnnotations updatedType)
+        {
+            Debug.Assert(underlyingLocal is object);
+            Debug.Assert(updatedContainingSymbol is object);
+            UnderlyingLocal = underlyingLocal;
+            ContainingSymbol = updatedContainingSymbol;
+            TypeWithAnnotations = updatedType;
+        }
+
+        private SourceLocalSymbol UnderlyingLocal { get; }
+        public override Symbol ContainingSymbol { get; }
+        public override TypeWithAnnotations TypeWithAnnotations { get; }
+
+        public override bool Equals(Symbol other, TypeCompareKind compareKind)
+        {
+            if (other == (object)this)
+            {
+                return true;
+            }
+
+            return other is LocalSymbol local
```

Comment:
>return other is LocalSymbol local [](start = 12, length = 33)

It is hard to tell if this is giving us correct behavior. I would separately handle a case when ```other``` is ```SourceLocalSymbol``` and when ```other``` is ```UpdatedContainingSymbolAndNullableAnnotationLocal```. For all other types the result should be false. Then, I would compare two ```SourceLocalSymbol``` for the exact match, and only then would compare type and container with the given ```compareKind```.